README.md
# url-shortify

URL shortener.

## Usage

## Development

```shell
# prepare environment
cp sample.env .env

# start supporting containers
docker compose up -d

# install dependencies
npm install

# apply SQL migrations
npm run migrations

# start web server
npm run start:dev

# run unit tests
npm run test:unit

# install pre-commit hooks
pre-commit install --hook-type commit-msg
```

Swagger UI is available at [http://127.0.0.1:3000/docs/static/index.html](http://127.0.0.1:3000/docs/static/index.html).  



CHANGELOG.md
# Changelog

## [1.1.1](https://github.com/url-shortify/url-shortify/compare/v1.1.0...v1.1.1) (2023-06-06)


### Bug Fixes

* remove unnecessary log output ([d41e61a](https://github.com/url-shortify/url-shortify/commit/d41e61ab6b811bf97f106581f8543aa14463f280))

## [1.1.0](https://github.com/url-shortify/url-shortify/compare/v1.0.0...v1.1.0) (2023-06-06)


### Features

* add config options when running behind proxy ([69d4a97](https://github.com/url-shortify/url-shortify/commit/69d4a9789052daa62f3fa8d09a31727b2539db7e))

## 1.0.0 (2023-06-06)


### Features

* init ([96a9783](https://github.com/url-shortify/url-shortify/commit/96a97833d5479665f7a9ffceed17302b01d25c6f))
* update pre-commit hooks ([8e24b05](https://github.com/url-shortify/url-shortify/commit/8e24b05669ee58d48a8f187c42259396a6247f00))


test/helper.js
const app = require('../src/app')

module.exports = async function (t) {
  const server = await app({
    APP_HOST: '127.0.0.1',
    APP_PORT: 3000,
    APP_LOGGER_LEVEL: 'silent',
  })

  t.teardown(server.close.bind(server))

  return server
}


test/health.test.js
const { test } = require('tap')
const pgmock2 = require('pgmock2')
const helper = require('./helper')

test('health route', async (t) => {
  const pg = new pgmock2.default()

  pg.add('SELECT NOW()', [], {
    rowCount: 1,
    rows: [{ now: '2023-05-10T14:19:38.979Z' }],
  })

  const server = await helper(t)
  server.pg = pg

  const res = await server.inject({
    url: '/health',
  })

  t.same(res.statusCode, 200)
  t.same(res.headers['cache-control'], 'no-store, max-age=0, private')
  t.same(res.headers['expires'], 0)
})


src/config.js
const envSchema = require('env-schema')

module.exports = envSchema({
  dotenv: true,
  schema: {
    type: 'object',
    required: ['PGHOST', 'PGPORT', 'PGDATABASE', 'PGUSER', 'PGPASSWORD'],
    properties: {
      PGHOST: {
        type: 'string',
      },
      PGPORT: {
        type: 'number',
      },
      PGDATABASE: {
        type: 'string',
      },
      PGUSER: {
        type: 'string',
      },
      PGPASSWORD: {
        type: 'string',
      },
      APP_HOST: {
        type: 'string',
        default: '127.0.0.1',
      },
      APP_PORT: {
        type: 'number',
        default: 3000,
      },
      APP_LOGGER_LEVEL: {
        type: 'string',
        default: 'info',
      },
      APP_TRUST_PROXY: {
        type: 'boolean',
        default: false,
      },
      APP_REQUEST_ID_HEADER: {
        type: 'string',
        default: 'request-id',
      },
      APP_DOMAIN: {
        type: 'string',
        default: '',
      },
      APP_HTTPS: {
        type: 'boolean',
        default: false,
      },
    },
  },
})


src/schemas.js
module.exports = {
  getHealth: {
    description: 'Check health',
    summary: 'Check health',
    tags: ['health'],
    response: {
      200: {
        description: 'Ok',
        type: 'object',
        properties: {},
      },
      '5xx': {
        description: 'Server error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
    },
  },
  getUrls: {
    description: 'Get list of short URLs',
    summary: 'Get list of short URLs',
    tags: ['url'],
    querystring: {
      type: 'object',
      properties: {
        offset: {
          type: 'integer',
          default: 0,
          minimum: 0,
          maximum: 50,
        },
      },
    },
    response: {
      200: {
        description: 'Ok',
        type: 'object',
        properties: {
          urls: {
            type: 'array',
            items: {
              type: 'object',
              required: ['id', 'shortUrl', 'targetUrl'],
              properties: {
                id: {
                  type: 'string',
                },
                shortUrl: {
                  type: 'string',
                },
                targetUrl: {
                  type: 'string',
                },
              },
            },
          },
        },
        examples: [
          {
            urls: [
              {
                id: 'o2fXhV',
                shortUrl: `https://example.com/o2fXhV`,
                targetUrl: 'https://www.example.com/support/disrupted-flights/latest-travel-updates/',
              },
            ],
          },
        ],
      },
      '4xx': {
        description: 'Client error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
      '5xx': {
        description: 'Server error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
    },
  },
  postUrl: {
    description: 'Create a short URL',
    summary: 'Create a short URL',
    tags: ['url'],
    body: {
      type: 'object',
      required: ['targetUrl'],
      properties: {
        targetUrl: {
          type: 'string',
          format: 'uri',
          pattern: '^https?://',
        },
      },
      examples: [{ targetUrl: 'https://www.example.com/support/disrupted-flights/latest-travel-updates/' }],
    },
    response: {
      200: {
        description: 'Ok',
        type: 'object',
        required: ['id', 'shortUrl', 'targetUrl'],
        properties: {
          id: {
            type: 'string',
          },
          shortUrl: {
            type: 'string',
          },
          targetUrl: {
            type: 'string',
          },
        },
        examples: [
          {
            id: 'o2fXhV',
            shortUrl: `https://example.com/o2fXhV`,
            targetUrl: 'https://www.example.com/support/disrupted-flights/latest-travel-updates/',
          },
        ],
      },
      '4xx': {
        description: 'Client error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
      '5xx': {
        description: 'Server error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
    },
  },
  getUrl: {
    description: 'Get short URL',
    summary: 'Get short URL',
    tags: ['url'],
    params: {
      type: 'object',
      required: ['id'],
      properties: {
        id: {
          type: 'string',
          pattern: `^[a-zA-Z0-9]{6}$`,
        },
      },
    },
    response: {
      200: {
        description: 'Ok',
        type: 'object',
        required: ['id', 'shortUrl', 'targetUrl'],
        properties: {
          id: {
            type: 'string',
          },
          shortUrl: {
            type: 'string',
          },
          targetUrl: {
            type: 'string',
          },
        },
        examples: [
          {
            id: 'o2fXhV',
            shortUrl: `https://example.com/o2fXhV`,
            targetUrl: 'https://www.example.com/support/disrupted-flights/latest-travel-updates/',
          },
        ],
      },
      '4xx': {
        description: 'Client error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
      '5xx': {
        description: 'Server error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
    },
  },
  deleteUrl: {
    description: 'Delete short URL',
    summary: 'Delete short URL',
    tags: ['url'],
    params: {
      type: 'object',
      required: ['id'],
      properties: {
        id: {
          type: 'string',
          pattern: '^[a-zA-Z0-9]{6}$',
        },
      },
    },
    response: {
      200: {
        description: 'Ok',
        type: 'object',
        properties: {},
      },
      '4xx': {
        description: 'Client error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
      '5xx': {
        description: 'Server error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
    },
  },
  redirectUrl: {
    description: 'Redirect short URL',
    summary: 'Redirect short URL',
    tags: ['url'],
    params: {
      type: 'object',
      required: ['id'],
      properties: {
        id: {
          type: 'string',
          pattern: `^[a-zA-Z0-9]{6}$`,
        },
      },
    },
    response: {
      301: {
        description: 'Redirect',
        type: 'object',
        headers: {
          Location: {
            type: 'string',
            description: 'Target URL',
          },
        },
      },
      '4xx': {
        description: 'Client error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
      '5xx': {
        description: 'Server error',
        type: 'object',
        properties: {
          statusCode: {
            type: 'integer',
          },
          error: {
            type: 'string',
          },
          message: {
            type: 'string',
          },
        },
      },
    },
  },
}


src/app.js
const fastify = require('fastify')
const fastifySensible = require('@fastify/sensible')
const fastifyHelmet = require('@fastify/helmet')
const fastifyPostgres = require('@fastify/postgres')
const fastifySwagger = require('@fastify/swagger')
const fastifySwaggerUI = require('@fastify/swagger-ui')
const Hashids = require('hashids/cjs')

const schemas = require('./schemas')
const packageJSON = require('../package.json')

module.exports = async function (config = {}) {
  const app = fastify({
    logger: {
      level: config.APP_LOGGER_LEVEL,
    },
    trustProxy: config.APP_TRUST_PROXY,
    requestIdHeader: config.APP_REQUEST_ID_HEADER,
  })

  app.decorate('config', config)
  app.decorate('utils', {
    hashids: new Hashids(packageJSON.name, 6),
    serverUrl: `${config.APP_HTTPS ? 'https' : 'http'}://${
      config.APP_DOMAIN ? config.APP_DOMAIN : `${config.APP_HOST}:${config.APP_PORT}`
    }`,
  })

  app.register(fastifySensible)
  app.register(fastifyHelmet)
  app.register(fastifyPostgres)

  await app.register(fastifySwagger, {
    openapi: {
      info: {
        title: packageJSON.name,
        description: packageJSON.description,
        version: packageJSON.version,
      },
      servers: [
        {
          url: app.utils.serverUrl,
        },
      ],
    },
    hideUntagged: true,
    exposeRoute: true,
  })

  app.register(fastifySwaggerUI, {
    routePrefix: '/docs',
    uiConfig: {
      docExpansion: 'list',
      deepLinking: false,
    },
    uiHooks: {
      onRequest: function (request, reply, next) {
        next()
      },
      preHandler: function (request, reply, next) {
        next()
      },
    },
    staticCSP: true,
    transformStaticCSP: (header) => header,
    transformSpecification: (swaggerObject) => {
      return swaggerObject
    },
    transformSpecificationClone: true,
  })

  app.get('/health', { schema: schemas.getHealth }, async function (request, reply) {
    reply.preventCache()
    await app.pg.query('SELECT NOW()')

    return {}
  })

  app.get('/urls', { schema: schemas.getUrls }, async function (request) {
    const result = await app.pg.query('SELECT * FROM urls LIMIT 50 OFFSET $1', [request.query.offset])

    return {
      urls: result.rows.map((row) => {
        return {
          id: row.hash_id,
          shortUrl: `${app.utils.serverUrl}/${row.hash_id}`,
          targetUrl: row.url,
        }
      }),
    }
  })

  app.post('/urls', { schema: schemas.postUrl }, async function (request) {
    const result = await app.pg.query('INSERT INTO urls(url) VALUES($1) RETURNING *', [request.body.targetUrl])
    const hashId = app.utils.hashids.encode(result.rows[0].id)
    await app.pg.query('UPDATE urls set hash_id = $1 WHERE id = $2', [hashId, result.rows[0].id])

    return {
      id: hashId,
      shortUrl: `${app.utils.serverUrl}/${hashId}`,
      targetUrl: result.rows[0].url,
    }
  })

  app.get('/urls/:id', { schema: schemas.getUrl }, async function (request, reply) {
    const id = app.utils.hashids.decode(request.params.id)[0]
    const result = await app.pg.query('SELECT * FROM urls WHERE id = $1', [id])

    if (result.rows.length === 0) {
      return reply.notFound()
    }

    return {
      id: result.rows[0].hash_id,
      shortUrl: `${app.utils.serverUrl}/${result.rows[0].hash_id}`,
      targetUrl: result.rows[0].url,
    }
  })

  app.delete('/urls/:id', { schema: schemas.deleteUrl }, async function (request, reply) {
    const id = app.utils.hashids.decode(request.params.id)[0]
    const result = await app.pg.query('SELECT * FROM urls WHERE id = $1', [id])

    if (result.rows.length === 0) {
      return reply.notFound()
    }

    await app.pg.query('DELETE FROM urls WHERE id = $1', [id])

    return {}
  })

  app.get('/:id', { schema: schemas.redirectUrl }, async function (request, reply) {
    const id = app.utils.hashids.decode(request.params.id)[0]
    const result = await app.pg.query('SELECT * FROM urls WHERE id = $1', [id])

    if (result.rows.length === 0) {
      return reply.notFound()
    }

    return reply.redirect(301, result.rows[0].url)
  })

  app.get('/', async function (request, reply) {
    return reply.redirect(302, '/docs')
  })

  await app.ready()
  app.swagger()

  return app
}
